### ISSUE TITLE
Widevine support in Arm64 Windows

### ISSUE THREAD

**Issue by mherrmann** [brave/brave-browser#28318]
On Arm64 Windows, 32-bit Brave supports Widevine. But Arm64 Brave does not. The goal of this issue is to fix this.

This issue is important because it is a blocker for full native Arm64 support on Windows. For example, it is a pre-requisite for #29112.

A good URL for testing whether a browser supports Widevine is https://bitmovin.com/demos/drm.

## Other browsers

Chrome doesn't offer a native Arm64 binary. Its 32-bit package supports Widevine.

Edge does offer a native Arm64 binary. It supports Widevine. Edge's installation directory contains a file called `widevinecdm_proxy.dll`. This gets loaded when a page with Widevine is opened in the browser. It's likely an Arm64 DLL that proxies the x86 Widevine implementation.

**Comment by mherrmann** [comment]
@marcpems mentioned Arm64EC as a potential option for loading the Widevine x64 DLL.

**Comment by marcpems** [comment]
Worth noting up front that Arm64EC is only MSVC today will produce that ABI although the Clang-cl tool has some work to support it. For linking it must be MSVC linker today.

Arm64EC will compile and link both intel and Arm instructions to use and x64 compatible ABI, so calling directly from Arm code to intel will just work within the same binary or across binaries. That has the potential to allow the widevine x64 dll to load and run within an Brave browser built for Arm64EC. Performance of EC is much better than X64 emulated because all system calls (from Arm or X64) are direct and not going through a shim. Arm64EC is considered a native ABI for Windows on Arm. 
https://www.youtube.com/watch?v=HmI6ip4o9as <- useful talk by Pedro on Arm64EC 

Would this be an option to explore for Brave?

**Comment by mihaiplesa** [comment]
@marcpems we'd need VS 2022 for that, right? We're currently on 2019.

**Comment by snickler** [comment]
> @marcpems we'd need VS 2022 for that, right? We're currently on 2019.

It will work with Visual Studio 2019, but for the best compatibility VS 2022 is the way to go (especially when targeting Windows 11). You should see `MSVC v142 -VS 2019 C++ ARM64EC build tools` in Individual Components from the installer. 

The Component ID for installing the workload is `Microsoft.VisualStudio.Component.VC.Tools.ARM64EC`. In Visual Studio 2019, it's still listed as experimental, but SHOULD work?

**Comment by mherrmann** [comment]
Do I understand correctly that with Arm64EC we'd compile the entire Brave browser with MS's toolchains instead of clang? If yes, that seems like a hefty price to pay just to get one DLL working.

**Comment by snickler** [comment]
> Do I understand correctly that with Arm64EC we'd compile the entire Brave browser with MS's toolchains instead of clang? If yes, that seems like a hefty price to pay just to get one DLL working.

See https://reviews.llvm.org/D134788

**Comment by d9j** [comment]
hope brave arm support will be stable soon
https://support.microsoft.com/en-us/windows/options-for-using-windows-11-with-mac-computers-with-apple-m1-and-m2-chips-cd15fd62-9b34-4b78-b0bc-121baa3c568c

**Comment by mherrmann** [comment]
Thank you @snickler. Could it be that LLD doesn't yet support Arm64EC? At least I'm getting `unknown machine: 42561` when trying to build Brave with it:

```
FAILED: obj/brave/build/rust/cxx_cppdeps.lib
..\..\third_party\llvm-build\Release+Asserts\bin\lld-link.exe /lib /OUT:obj/brave/build/rust/cxx_cppdeps.lib /nologo /ignore:4221 /llvmlibthin /WX @obj/brave/build/rust/cxx_cppdeps.lib.rsp
obj/brave/build/rust/cxx_cppdeps/remap_alloc.obj: unknown machine: 42561
```

**Comment by snickler** [comment]
> Thank you @snickler. Could it be that LLD doesn't yet support Arm64EC? At least I'm getting `unknown machine: 42561` when trying to build Brave with it:
> 
> ```
> FAILED: obj/brave/build/rust/cxx_cppdeps.lib
> ..\..\third_party\llvm-build\Release+Asserts\bin\lld-link.exe /lib /OUT:obj/brave/build/rust/cxx_cppdeps.lib /nologo /ignore:4221 /llvmlibthin /WX @obj/brave/build/rust/cxx_cppdeps.lib.rsp
> obj/brave/build/rust/cxx_cppdeps/remap_alloc.obj: unknown machine: 42561
> ```

Hmmm which triple are you using? Also, I wonder if this is more of a Rust issue?

**Comment by mherrmann** [comment]
> which triple are you using?

`arm64-pc-windows-msvc19.29.0`. (I'm still on VS 2019.)

**Comment by snickler** [comment]
> > which triple are you using?
> 
> `arm64ec-pc-windows-msvc19.29.0`. (I'm still on VS 2019.)

I'm assuming it also causes issues if you strip off the msvc version?

**Comment by mherrmann** [comment]
Yes. I need to check whether it's even possible to compile a simple Hello World program to Arm64EC with llvm.

**Comment by mherrmann** [comment]
It appears to me that at least parts of llvm don't yet support Arm64EC. The machine type `42561` in the error above is `A641` in Hex, which is [`IMAGE_FILE_MACHINE_ARM64EC`](https://github.com/llvm/llvm-project/blob/72e8b286f03c7f6bacbec10ca9883f77d482284c/llvm/include/llvm/BinaryFormat/COFF.h#L101). The error message comes from https://github.com/llvm/llvm-project/blob/72e8b286f03c7f6bacbec10ca9883f77d482284c/llvm/lib/ToolDrivers/llvm-lib/LibDriver.cpp#L173, which clearly has no `case` for that constant.

**Comment by snickler** [comment]
> It appears to me that at least parts of llvm don't yet support Arm64EC. The machine type `42561` in the error above is `A641` in Hex, which is [`IMAGE_FILE_MACHINE_ARM64EC`](https://github.com/llvm/llvm-project/blob/72e8b286f03c7f6bacbec10ca9883f77d482284c/llvm/include/llvm/BinaryFormat/COFF.h#L101). The error message comes from https://github.com/llvm/llvm-project/blob/72e8b286f03c7f6bacbec10ca9883f77d482284c/llvm/lib/ToolDrivers/llvm-lib/LibDriver.cpp#L173, which clearly refuses to handle that case.

Well Darn, it DOES seem that's the case for llvm-lib. 

There's a PR for it, which makes me wonder just *which* parts of llvm support targeting arm64ec since there's many references to it in the source, along with the referenced urls showing such. 

https://reviews.llvm.org/D143541

**Comment by mherrmann** [comment]
Yes, that PR seems to add the constant precisely to the place I linked to. I wonder whether they will merge it soon, whether that is the only change that's missing, and then how stable/production ready it will be.

**Comment by nsait-linaro** [comment]
Qualcomm is working on adding ARM64EC support to LLVM.  

https://discourse.llvm.org/t/initial-patches-for-arm64ec-windows-11-now-posted/62449

I've reached out to Qualcomm and they have asked if we have any build recipes for Brave available so they can try to validate the implementation.

**Comment by mherrmann** [comment]
@nsait-linaro yes - follow the instructions at https://github.com/brave/brave-browser/ and additionally install `MSVC v142 -VS 2019 C++ ARM64EC build tools` in Visual Studio 2019. Before you do `npm run build`, replace `arm64-windows` in 2-3 .gn files by `arm64ec-windows` in `src/`. Then do `npm run build -- --target_os=arm64`.

**Comment by mherrmann** [comment]
[Comment](https://www.youtube.com/watch?v=HmI6ip4o9as&lc=UgzHya-HYRc-8JpDZCN4AaABAg.9nZAB23muff9nZgelOTJFO) from Pedro Justo about how Edge does it with Widevine:

> You can't do that in the same process. The x64 DLL can't load in the Arm64 Classic process - period. Your "proxy" must involve another process and use IPC. In the Arm64 process load an Arm64 DLL + Create an Arm64EC EXE to load your x64 DLL + use IPC mechanisms like RPC or DCOM to communicate between them. This is what Arm64 Edge does with Widevine, for example.

**Comment by mherrmann** [comment]
There seems to be movement on the Chromium side for Widevine Windows Arm64 support: https://github.com/chromium/chromium/commit/77efb41adb7b9a99f0827926027ca0f34c381544

**Comment by snickler** [comment]
> There seems to be movement on the Chromium side for Widevine Windows Arm64 support: [chromium/chromium@77efb41](https://github.com/chromium/chromium/commit/77efb41adb7b9a99f0827926027ca0f34c381544)

And this is the latest widevine binary for win-arm64!

https://dl.google.com/widevine-cdm/4.10.2557.0-win-arm64.zip

**Comment by mherrmann** [comment]
That's great! And I just tested that the binary can in fact be loaded into Arm64 Brave :)

**Comment by mherrmann** [comment]
@nsait-linaro

> this probably means we are not blocked on ARM64EC support for LLVM?

As a general statement, I'd say this is correct. More specifically, we no longer need Arm64EC for Widevine. We may still decide to use it for on-demand updates (see #28380). But I'd say that's unlikely.

> And https://github.com/brave/brave-browser/issues/29112 is no longer relevant

That's not correct. #29112 is the fix Microsoft are using for #28380, and I think we should follow it.

**Comment by nsait-linaro** [comment]
Sorry I meant the windows update issue with github.com/google/omaha are no longer relevant as we can always install arm64 binaries.

> 
> That's not correct. #29112 is the fix Microsoft are using for #28380, and I think we should follow it.

**Comment by mherrmann** [comment]
Yes. :)

**Comment by MadhaviSeelam** [comment]
Verification `In progress` using
```
Brave | 1.59.86 Chromium: 117.0.5938.62 (Official Build) beta (arm64)
-- | --
Revision | 3c4c1c3b2eef0fcbe8485c67e1df504942deadca
OS | Windows 11 Version 22H2 (Build 22621.2283)
```

1. Installed 1.59.86
2. launched Brave
3. visit https://bitmovin.com/demos/drm in a new tab page
4. confirmed I got prompted to Install and run Google Widevine
5. clicked Allow
6. played the video for protected DRM stream
7. opened brave://components
8. confirmed `Widevine Content Decryption Module`  is loaded


step 4 | step 6
---- | ----  
<img width="953" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/b9ad3c40-10bf-48fb-b681-fcdabb2bf9a5">|<img width="953" alt="image" src="https://github.com/brave/brave-browser/assets/98358127/1352a2b6-c196-44be-81ab-8177b669bbb6">


brave://components | WidevineCdm folder
------ | -------
![image](https://github.com/brave/brave-browser/assets/98358127/46cb821a-aca0-4b07-9323-aa247f8471e0)
![image](https://github.com/brave/brave-browser/assets/98358127/d0eb1244-2126-4f90-a535-d147b177896a)
